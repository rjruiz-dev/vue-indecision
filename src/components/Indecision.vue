<template>
    <!-- <img src="https://via.placeholder.com/250" alt="bg"> -->
    <!-- v-bind, utilizo shortcuod :src -->
    <!-- v-if: como la img es null = false  -->
    <img  v-if="img" :src="img" alt="bg">

    <div class="bg-dark">

    </div>
               
    <div class="indecision-container">
        <input
          type="text"
          placeholder="Hazme una pregunta"
          v-model="question"> <!-- enlazar el input con la propiedad reactiva --> 
          
        <p>Recuerda terminar con un signo de interrogación (?)</p>

        <!-- renderizo este segmento  so la pregunta es valida -->
        <div v-if="isValidQuestion">
            <h2>{{ question }}</h2>
            <h1>{{ answer }}</h1>
        </div>

    </div>
</template>

<script>
export default {
     data() {
        return {
            // propiedad reactiva
            question: null,
            answer: null,
            img: null,
            // es una bandera para saber cunado debo mostrar cierto texto
            isValidQuestion: false
        }
    },
    methods: {
        async getAnswer() {

            try {
                // cuando se esta haciendo la peticion el msj: Pensando...
                this.answer = 'Pensando...'

                // peticion con fetch
                // es una promesa, extraer el json
                // const data = await fetch('https://yesno.wtf/api').then( r => r.json() )
                // console.log(data);
                
                // desectructuracion de lo q me intersa
                const { answer, image } = await fetch('https://yesno.wtf/api').then( r => r.json() )
                // console.log(answer);

                // es la propiedad = la const desectructurada
                this.answer = answer === 'yes' ? 'Si!' : 'No!'
                this.img = image 

            } catch (error) {
                //  si el fetch falla
                console.log('IndecisionComponent: ', error);
                this.answer = 'No se pudo cargar del API'
                this.img = null
            }
            
        }
    },    
    // watch: como si fuera un obj, en este caso esta pendiente de la pro reactiva question
    watch: {
        // cuando el question cambia
        question (value, oldValue){
            // loqueamos el obj {}
            // console.log({value, oldValue });

            this.isValidQuestion = false

            console.log( { value });

            // si el nuevo valor incluye singo ?
            // console.log(value.includes('?')); // devuelve true o false
            if ( !value.includes('?') ) return

            this.isValidQuestion = true

            // TODO: Realizar petición http invocando el metodo getAnswer()
            this.getAnswer()
        }
    }

}
</script>

<style >

    img, .bg-dark {
        height: 100vh;
        left: 0px;
        max-height: 100%;
        max-width: 100%;
        position: fixed;
        top: 0px;
        width: 100vw;
    }

    .bg-dark {
        background-color: rgba(0, 0, 0, 0.4);
    }

    .indecision-container {
        position: relative;
        z-index: 99;
    }
    
    input {
        width: 250px;
        padding: 10px 15px;
        border-radius: 5px;
        border: none;
    }
    input:focus {
        outline: none;
    }

    p {
        color: white;
        font-size: 20px;
        margin-top: 0px;
    }

    h1, h2 {
        color: white;
    }
    
    h2 {
        margin-top: 150px;
    }

</style>