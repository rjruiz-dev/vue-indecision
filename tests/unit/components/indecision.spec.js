import { shallowMount} from '@vue/test-utils'
import Indecision from '@/components/Indecision'

describe('Indecision Component', () => {
    let wrapper 
    let clgSpy // console.log espia

    // mock del fetch
    global.fetch = jest.fn( () => Promise.resolve({
        // respuesta de la promesa
        json: () => Promise.resolve({
            // resuleve la petision del endpoin retorna
           answer: 'yes',
           forced: false,
           image: 'https://yesno.wtf/assets/yes/2.gif'
        })
    }) ) //funcion jest Promise.resolve: crea promesa y la resuelve de manera instantanea

    beforeEach(() => {
        wrapper = shallowMount(Indecision)

        clgSpy = jest.spyOn( console, 'log') // obj, metodo

        // proceso de limpieza individual para escribir el simbolo de "?"...
        // clgSpy.clear

        // proceso de limpieza para todo el test suite
        jest.clearAllMocks()
    })

    test('debe hacer match con el snapshot', () => {
        expect( wrapper.html() ).toMatchSnapshot() // captura del codigo:
    })

    test('escribir en el input no debe de disparar nada (console.log)', async () => {
        
        // funcion espia
        const getAnswerSpy = jest.spyOn( wrapper.vm, 'getAnswer') //  wrapper instancia del componente donde esta montado, vm intnacia de vue


        // hacer la referencia al input
        const input = wrapper.find('input') 
        
        // si la modificiacion afecta al doom, hay que esperar a q se vuelva a renderiazar por eso poner await
        await input.setValue('Hola Mundo')

        // evaluar si el console.log fue disparado toHaveBeenCalled
        // expect( clgSpy ).toHaveBeenCalled()
        expect( clgSpy ).toHaveBeenCalledTimes(1)
        
        // getAnswer no debe ser llamado
        expect( getAnswerSpy ).toHaveBeenCalledTimes(0)
        expect( getAnswerSpy ).not.toHaveBeenCalled()

    })

    test('escribir el simbolo de "?" debe de disparar el getAnswer', async () => {
        // funcion espia
        const getAnswerSpy = jest.spyOn( wrapper.vm, 'getAnswer') //  wrapper instancia del componente donde esta montado, vm intnacia de vue

        // hacer la referencia al input
        const input = wrapper.find('input') 
        
        // si la modificiacion afecta al doom, hay que esperar a q se vuelva a renderiazar por eso poner await
        await input.setValue('?')

        // getAnswer debe ser llamado
        expect( getAnswerSpy ).toHaveBeenCalledTimes(1)
        expect( getAnswerSpy ).toHaveBeenCalled()
    })

    test('prueba en getAnswer', async () => {
        // obtener un si
        await wrapper.vm.getAnswer()

        // console.log(wrapper.vm.img);
        // console.log(wrapper.vm.answer);

        // esperando que la imgen exista
        const img = wrapper.find('img')

        expect( img.exists()).toBeTruthy()
        expect( wrapper.vm.img).toBe('https://yesno.wtf/assets/yes/2.gif') // que se la misma imagen que estoy esperando
        expect( wrapper.vm.answer).toBe('Si!')
    })

    test('pruebas en getAnswer - fallo en el API', async () => {
        // simular el fallo en el api 
        fetch.mockImplementationOnce( () => Promise.reject('API is down'))
       
        await wrapper.vm.getAnswer()

        // si todo sale mal por eje la imagen no existe
        const img = wrapper.find('img')

        expect( img.exists() ).toBeFalsy()
        expect( wrapper.vm.answer).toBe('No se pudo cargar del API')
    })
})