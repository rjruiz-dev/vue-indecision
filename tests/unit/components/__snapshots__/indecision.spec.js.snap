// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Component debe hacer match con el snapshot 1`] = `
<!-- <img src="https://via.placeholder.com/250" alt="bg"> -->
<!-- v-bind, utilizo shortcuod :src -->
<!-- v-if: como la img es null = false  -->
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Hazme una pregunta"><!-- enlazar el input con la propiedad reactiva -->
  <p>Recuerda terminar con un signo de interrogaci√≥n (?)</p><!-- renderizo este segmento  so la pregunta es valida -->
  <!--v-if-->
</div>
`;
